Add initial integration of iterative scheduling (#88)
triton-inference-server/tutorials: Conceptual_Guide/Part_7-iterative_scheduling/README.md

#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NVIDIA CORPORATION nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

# Deploying a GPT-2 Model using Python Backend and Iterative Scheduling

In this tutorial, we will deploy a GPT-2 model using the Python backend and

> [iterative scheduling](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/model_configuration.html#iterative-sequences).

Reviewer: ```suggestion
demonstrate the [iterative scheduling](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/model_configuration.html#iterative-sequences) feature.
```

## Prerequisites

Before getting started with this tutorial, make sure you're familiar
with the following concepts:

* [Python Backend](https://github.com/triton-inference-server/python_backend)

> * [Triton-Server Quick Start](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/getting_started/quickstart.html)

Reviewer: nit: I'd swap the order of these two, in case a user is familiar with neither - they should do the quickstart first.

## Iterative Scheduling

Iterative scheduling is a technique that allows the Triton Inference Server to
schedule the same request multiple times with the same input. This is useful for
models that have an auto-regressive loop. Iterative scheduling enables Triton
Server to implement inflight batching for your models and gives you the ability
to combine new sequences as they are arriving with inflight sequences. Please
take a look at the [Orca](https://www.usenix.org/system/files/osdi22-yu.pdf)
paper for more details.

## Tutorial Overview

In this tutorial we deploy two models:

* simple-gpt2: This model receives a batch of requests and proceeds to the next
batch only when it is done generating tokens for the current batch.

> * iterative-gpt2: This model uses iterative scheduling and is able to process
> new sequences in the batch even when it is still generating tokens for the
> previous sequences.

Reviewer: ```suggestion
* iterative-gpt2: This model uses iterative scheduling to process
new sequences in a batch even when it is still generating tokens for the
previous sequences.
```

```

* Next, install all the dependencies required by the models running in the
python backend and login with your [huggingface token](https://huggingface.co/settings/tokens)
(Account on [HuggingFace](https://huggingface.co/) is required).

```
pip install transformers[torch]
huggingface-cli login
```

Then, start the server:

```
tritonserver --model-repository=/models
```

### Step 2: Install the client side dependencies

In another terminal install the client dependencies:

```
pip3 install tritonclient[grpc]
pip3 install prompt_toolkit
```

### Step 3: Run the client

The simple-gpt2 model doesn't use iterative scheduling and will proceed to the
next batch only when it is done generating tokens for the current batch.

Run the following command to start the client:

```
python3 client/client.py --model simple-gpt2
```

As you can see, the tokens for one request are processed first before proceeding
to the next request.

Run `Ctrl+C` to stop the client.

The iterative scheduler is able to incorporate new requests as they are arriving
in the server.

Run the following command to start the client:
```
python3 client/client.py --model iterative-gpt2
```

> As you can see, the tokens for both prompts are getting generated simultaneously.

Reviewer 2: I think the demo should be mentioned beforehand because "as you can see" is referring to the output shown in the demo video. In the current structure the reader would expect the output is inlined in this section.

### Demo

> [![asciicast](https://asciinema.org/a/WeDlFwRuOip6q7EgQMZmWKRE7.svg)](https://asciinema.org/a/WeDlFwRuOip6q7EgQMZmWKRE7)

Reviewer: Nice inclusion of inline recording! Do you know how long this recording link will persist?

Author: I couldn't find anything about how long they'll keep the recording. I've been using it for past 5 years and have not noticed removal of any recording.

Reviewer: Sweet :rocket:

## Iterative Scheduling

Iterative scheduling is a technique that allows the Triton Inference Server to
schedule the same request multiple times with the same input. This is useful for
models that have an auto-regressive loop. Iterative scheduling enables Triton
Server to implement inflight batching for your models and gives you the ability
to combine new sequences as they are arriving with inflight sequences. Please
take a look at the [Orca](https://www.usenix.org/system/files/osdi22-yu.pdf)
paper for more details.

## Tutorial Overview

In this tutorial we deploy two models:

* simple-gpt2: This model receives a batch of requests and proceeds to the next
batch only when it is done generating tokens for the current batch.

* iterative-gpt2: This model uses iterative scheduling to process
new sequences in a batch even when it is still generating tokens for the
previous sequences

### Demo

[![asciicast](https://asciinema.org/a/WeDlFwRuOip6q7EgQMZmWKRE7.svg)](https://asciinema.org/a/WeDlFwRuOip6q7EgQMZmWKRE7)

### Step 1: Prepare the Server Environment

* First, run the Triton Inference Server Container:

```
# Replace yy.mm with year and month of release.

> docker run --gpus=all -it --shm-size=256m --rm -p8000:8000 -p8001:8001 -p8002:8002 -v ${PWD}:/workspace/ -v ${PWD}/model_repository:/models nvcr.io/nvidia/tritonserver:yy.mm-py3 bash

Reviewer 3: do we need a note here on the earliest version this will work on?

```

* Next, install all the dependencies required by the models running in the
python backend and login with your [huggingface token](https://huggingface.co/settings/tokens)
(Account on [HuggingFace](https://huggingface.co/) is required).

```

> pip install transformers[torch]

Reviewer 3: I'm almost in favor of writing a small docker file or instructions to commit and save the image for reuse.

It's a little cumbersome to have to install everytime if you wanted to try out different things .... any thoughts

Author: The user can run `docker commit ...` if they want to save their changes and re-use the image.

Reviewer 3: yeah let's see if we can add a one liner to have docker commit to update the image -

> huggingface-cli login

Reviewer 3: I *think* login is not required for this model - we could remove it / consider it optional

feature.

## Prerequisites

Before getting started with this tutorial, make sure you're familiar
with the following concepts:

* [Triton-Server Quick Start](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/getting_started/quickstart.html)
* [Python Backend](https://github.com/triton-inference-server/python_backend)

## Iterative Scheduling

Iterative scheduling is a technique that allows the Triton Inference Server to
schedule the same request multiple times with the same input. This is useful for
models that have an auto-regressive loop. Iterative scheduling enables Triton
Server to implement inflight batching for your models and gives you the ability
to combine new sequences as they are arriving with inflight sequences.

## Tutorial Overview

In this tutorial we deploy two models:

* simple-gpt2: This model receives a batch of requests and proceeds to the next
batch only when it is done generating tokens for the current batch.

* iterative-gpt2: This model uses iterative scheduling to process
new sequences in a batch even when it is still generating tokens for the
previous sequences

### Demo

> [![asciicast](https://asciinema.org/a/WeDlFwRuOip6q7EgQMZmWKRE7.svg)](https://asciinema.org/a/WeDlFwRuOip6q7EgQMZmWKRE7)

Reviewer: Does the recording need to change with `tqdm` changes?
